accounts/accounts.go

```diff
@@ -0,0 +1,178 @@
+}
+
+func (acc *Account) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + acc.Address.Hex() + `"`), nil
+}
+
+func (acc *Account) UnmarshalJSON(raw []byte) error {
+	return json.Unmarshal(raw, &acc.Address)
+}
+
+// Manager is an overarching account manager that can communicate with various
+// backends for signing transactions.
+type Manager struct {
+	backends []Backend                // List of currently registered backends (ordered)
+	index    map[reflect.Type]Backend // Set of currently registered backends
+	lock     sync.RWMutex
+}
+
+// NewManager creates a generic account manager to sign transaction via various
+// supported backends.
+func NewManager(backends ...Backend) *Manager {
+	am := &Manager{
+		backends: backends,
+		index:    make(map[reflect.Type]Backend),
+	}
+	for _, backend := range backends {
+		am.index[reflect.TypeOf(backend)] = backend
+	}
+	return am
+}
+
+// Backend retrieves the backend with the given type from the account manager.
+func (am *Manager) Backend(backend reflect.Type) Backend {
+	return am.index[backend]
+}
+
+// Accounts returns all signer accounts registered under this account manager.
+func (am *Manager) Accounts() []Account {
+	am.lock.RLock()
+	defer am.lock.RUnlock()
+
+	var all []Account
+	for _, backend := range am.backends { // TODO(karalabe): cache these after subscriptions are in
+		accounts := backend.Accounts()
+		for i := 0; i < len(accounts); i++ {
+			accounts[i].backend = backend
+		}
+		all = append(all, accounts...)
+	}
+	return all
+}
+
+// HasAddress reports whether a key with the given address is present.
+func (am *Manager) HasAddress(addr common.Address) bool {
+	am.lock.RLock()
+	defer am.lock.RUnlock()
+
+	for _, backend := range am.backends {
+		if backend.HasAddress(addr) {
+			return true
+		}
+	}
+	return false
+}
+
+// SignHash requests the account manager to get the hash signed with an arbitrary
+// signing backend holding the authorization for the specified account.
+func (am *Manager) SignHash(acc Account, hash []byte) ([]byte, error) {
+	am.lock.RLock()
+	defer am.lock.RUnlock()
+
+	if err := am.ensureBackend(acc); err != nil {
+		return nil, err
+	}
+	return acc.backend.SignHash(acc, hash)
+}
+
+// SignTx requests the account manager to get the transaction signed with an
+// arbitrary signing backend holding the authorization for the specified account.
+func (am *Manager) SignTx(acc Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
+	am.lock.RLock()
+	defer am.lock.RUnlock()
+
+	if err := am.ensureBackend(acc); err != nil {
+		return nil, err
+	}
+	return acc.backend.SignTx(acc, tx, chainID)
+}
+
+// SignHashWithPassphrase requests the account manager to get the hash signed with
+// an arbitrary signing backend holding the authorization for the specified account.
+func (am *Manager) SignHashWithPassphrase(acc Account, passphrase string, hash []byte) ([]byte, error) {
```

Comment:
This is really a way to specify an extra field of information to the backend for signing which it either uses or not (hw wallets won't use it, keystores will). I couldn't figure out a cleaner way to support this operation without requiring user input callbacks.